### PR TITLE
Monitoring of the GitHub client, of the GitHub API Rate Limit remaini…

### PR DESCRIPTION
**Pull request by cyrille-leclerc**
Monitoring of the GitHub client, of the GitHub API Rate Limit remaining requests as a gauge with the following attributes:
* `enduser.id`: login of the authenticated GitHub user (e.g. `john-doe`)
* `github.api.url`: e.g. `https://api.github.com`
* `github.app.id`: e.g. `1234567`
* `github.app.name`: e.g. `My Jenkins GitHub App`
* `github.app.owner`: e.g. `cyrille-leclerc`. Maybe be `null`
* TODO @cyrille-leclerc to document addition attributes

Implementation of 
* https://github.com/jenkinsci/opentelemetry-plugin/issues/411

Tested with
* Jenkins current LTS 2.332.2 and GitHub Branch Source plugin [1598.v91207e9f9b_4a_](https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/1598.v91207e9f9b_4a_)
* Jenkins 2.277.1 and GitHub Branch Source plugin [2.10.2](https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/github-branch-source-2.10.2)

Looks like on the Prometheus Exposition Format web page of the Otel Collector:

### GitHub User Authentication

```
jenkins_github_api_rate_limit_remaining_requests
    {
    enduser_id="cyrille-leclerc"
    github_api_url="https://api.github.com",
    github_authentication="login:cyrille-leclerc",
    ...} 
    4982
```

### GitHub App Authentication

```
github_api_rate_limit_remaining_requests
    {
    github_api_url="https://api.github.com",
    github_app_id=194536,
    github_app_name="My Jenkins GitHub App",
    github_app_owner="My Jenkins GitHub App",
    github_authentication="application:id=194536,name="My Jenkins GitHub App",owner="My Jenkins GitHub App""
    ...} 
    4982
```

### Anonymous Authentication
```
github_api_rate_limit_remaining_requests
    {
    github_api_url="https://api.github.com",
    github_authentication="anonymous"
    ...} 
    41
```

### Details

```
# HELP github_api_rate_limit_remaining_requests GitHub Repository API rate limit remaining requests
# TYPE github_api_rate_limit_remaining_requests gauge

github_api_rate_limit_remaining_requests{github_api_url="https://api.github.com",github_authentication="anonymous",host_arch="x86_64",host_name="MacBook-Pro.localdomain",jenkins_opentelemetry_plugin_version="2.3.1-SNAPSHOT (private-6de0bf3a-cyrilleleclerc)",jenkins_url="http://localhost:8080/jenkins/",jenkins_version="2.277.1",os_description="Mac OS X 11.3",os_type="darwin",process_runtime_description="AdoptOpenJDK OpenJDK 64-Bit Server VM 11.0.11+9",process_runtime_name="OpenJDK Runtime Environment",process_runtime_version="11.0.11+9",service_name="jenkins",service_namespace="jenkins",service_version="2.277.1",telemetry_sdk_language="java",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="1.13.0"} 58
github_api_rate_limit_remaining_requests{enduser_id="cyrille-leclerc",github_api_url="https://api.github.com",github_authentication="login:cyrille-leclerc",host_arch="x86_64",host_name="MacBook-Pro.localdomain",jenkins_opentelemetry_plugin_version="2.3.1-SNAPSHOT (private-6de0bf3a-cyrilleleclerc)",jenkins_url="http://localhost:8080/jenkins/",jenkins_version="2.277.1",os_description="Mac OS X 11.3",os_type="darwin",process_runtime_description="AdoptOpenJDK OpenJDK 64-Bit Server VM 11.0.11+9",process_runtime_name="OpenJDK Runtime Environment",process_runtime_version="11.0.11+9",service_name="jenkins",service_namespace="jenkins",service_version="2.277.1",telemetry_sdk_language="java",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="1.13.0"} 4973
github_api_rate_limit_remaining_requests{github_api_url="https://api.github.com",github_app_id="194536",github_app_name="Jenkins mvn hpi:run",github_app_owner="Jenkins mvn hpi:run",github_authentication="app:id=194536,app.name=\"Jenkins mvn hpi:run\",app.owner=Jenkins mvn hpi:run",host_arch="x86_64",host_name="MacBook-Pro.localdomain",jenkins_opentelemetry_plugin_version="2.3.1-SNAPSHOT (private-6de0bf3a-cyrilleleclerc)",jenkins_url="http://localhost:8080/jenkins/",jenkins_version="2.277.1",os_description="Mac OS X 11.3",os_type="darwin",process_runtime_description="AdoptOpenJDK OpenJDK 64-Bit Server VM 11.0.11+9",process_runtime_name="OpenJDK Runtime Environment",process_runtime_version="11.0.11+9",service_name="jenkins",service_namespace="jenkins",service_version="2.277.1",telemetry_sdk_language="java",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="1.13.0"} 12473
```
# Noteworthy

* Uses a "hack" with Java reflection API to access the metrics on the GitHub client
* doesn't verify the Rate Limit at each time interval, just uses the "last requested Rate Limit"

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by cyrille-leclerc** [comment]
I forgot to add on the metric an attribute with the credentials identifier

**Comment by cyrille-leclerc** [comment]
@timja can you test this feature?

**Comment by timja** [comment]
I’ll test it ~tomorrow, thanks!

**Comment by timja** [comment]
This doesn't work on newer GitHub API, picked up in tests in this PR which updates the bom version:
https://github.com/jenkinsci/opentelemetry-plugin/pull/414

File: src/main/java/io/jenkins/plugins/opentelemetry/init/GitHubClientMonitoring.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                .setDescription("GitHub Repository API rate limit remaining requests")
+                .setUnit("1")
+                .buildWithCallback(gauge -> {
+                    logger.log(Level.FINE, () -> "Collect GitHub client API rate limit metrics");
+                    try {
+                        Map<GitHub, ?> reverseLookup = (Map<GitHub, ?>) connectorReverseLookupField.get(null);
+                        reverseLookup.keySet().forEach(gitHub -> {
+                            GHRateLimit ghRateLimit = gitHub.lastRateLimit();
+                            try {
+                                String login = gitHub.isAnonymous() ? "#annonymous#" : gitHub.getMyself().getLogin();
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/opentelemetry/init/GitHubClientMonitoring.java:64]:
```suggestion
                                String login = gitHub.isAnonymous() ? "#anonymous#" : gitHub.getMyself().getLogin();
```

**Comment by cyrille-leclerc** [comment]
@timja can you please test?

**Pull request merged**